src/ReductionMsg.chpl

```diff
@@ -625,30 +621,100 @@ module ReductionMsg
     proc segAny(values:[] bool, segments:[?D] int): [D] bool {
       var res: [D] bool;
       if (D.size == 0) { return res; }
-      forall (r, low, i) in zip(res, segments, D) {
-        var high: int;
-        if (i < D.high) {
-          high = segments[i+1] - 1;
+      const sums = segSum(values, segments);
+      res = (sums > 0);
+      return res;
+    }
+
+    proc segAll(values:[] bool, segments:[?D] int): [D] bool {
+      var res: [D] bool;
+      if (D.size == 0) { return res; }
+      const sums = segSum(values, segments);
+      const lengths = segCount(segments, values.domain.high + 1);
+      res = (sums == lengths);
+      return res;
+    }
+
+    proc segOr(values:[?vD] int, segments:[?D] int): [D] int {
+      // Bitwise OR does not have an inverse, so this cannot be
+      // done with a scan. Each segment's values must be reduced
+      // separately.
+      var res: [D] int;
+      if (D.size == 0) { return res; }
+      forall (i, s, r) in zip(D, segments, res) {
+        // Find segment end
+        var e: int;
+        if i < D.high {
+          e = segments[i+1];
         } else {
-          high = values.domain.high;
+          e = vD.high;
+        }
+        // Run computation on locale where segment values start
+        // At most, numLocales-1 segments will have to get remote values
+        // Some results will have to be sent to remote locales
+        ref start = values[s];
+        const startLocale = start.locale.id;
+        const hereLocale = here.id;
+        if startLocale == hereLocale {
+          r = | reduce values[s..e];
+        } else {
+          on startLocale {
+            r = | reduce values[s..e];
+          }
```

Comment:
I just happened to notice this from the excerpt in the OP:  I wouldn't expect this type of specialization for locality in the event that the on-clause refers to the current locale to be necessary.  The Chapel compiler should already be doing this kind of optimization automatically for every on-clause.  If you find that this kind of manual specialization results in an improvement, that's something we'd definitely want to know about.